initialize/user_data.go

```diff
@@ -15,40 +15,407 @@
+			}
 		}
+		switch partHeader.mediaType {
+		case "text/cloud-config":
+			part, err := payloadAsCloudConfigPart(partHeader.fileName, string(body), env)
+			if err != nil {
+				return []UserDataPart{}, fmt.Errorf("error parsing cloud-config: %w", err)
+			}
+
+			udParts = append(udParts, part)
+			continue
+		case "text/x-shellscript":
+			part, err := payloadAsScriptPart(partHeader.fileName, string(body), env)
+			if err != nil {
+				return []UserDataPart{}, fmt.Errorf("error parsing script: %w", err)
+			}
+			udParts = append(udParts, part)
+			continue
+		case "application/gzip",
+			"application/gzip-compressed",
+			"application/gzipped",
+			"application/x-compress",
+			"application/x-compressed",
+			"application/x-gunzip",
+			"application/x-gzip",
+			"application/x-gzip-compressed":
 
-		if err := cc.Decode(); err != nil {
-			return nil, err
+			gzr, err := gzip.NewReader(bytes.NewReader(body))
+			if err != nil {
+				return []UserDataPart{}, fmt.Errorf("error reading gzip: %w", err)
+			}
+			body, err = ioutil.ReadAll(gzr)
+			if err != nil {
+				gzr.Close()
+				return []UserDataPart{}, fmt.Errorf("error reading gzip: %w", err)
+			}
+			gzr.Close()
+			// with the gzip wrapper removed, we can now parse the part. Fallthrough to
+			// the default condition, which will attempt to detect the part type and return
+			// a UserDataPart.
+			fallthrough
+		default:
+			parsedParts, err := partsFromUserData(string(body), env)
+			if err != nil {
+				return []UserDataPart{}, fmt.Errorf("error parsing part: %w", err)
+			}
+			udParts = append(udParts, parsedParts...)
 		}
+	}
+
+	return udParts, nil
+}
+
+func payloadAsScriptPart(name, payload string, env *Environment) (UserDataPart, error) {
+	if env == nil {
+		return UserDataPart{}, fmt.Errorf("environment is nil")
+	}
+
+	userdata := env.Apply(payload)
+	if !config.IsScript(userdata) {
+		return UserDataPart{}, fmt.Errorf("payload is not a script")
+	}
+	script, err := config.NewScript(userdata)
+	if err != nil {
+		return UserDataPart{}, err
+	}
+	return UserDataPart{
+		userDataType: ScriptType,
+		contents:     userdata,
+		script:       script,
+		fileName:     name,
+	}, nil
+}
+
+func payloadAsCloudConfigPart(name, payload string, env *Environment) (UserDataPart, error) {
+	if env == nil {
+		return UserDataPart{}, fmt.Errorf("environment is nil")
+	}
+
+	userdata := env.Apply(payload)
+	cc, err := config.NewCloudConfig(userdata)
+	if err != nil {
+		return UserDataPart{}, err
+	}
+
+	if err := cc.Decode(); err != nil {
+		return UserDataPart{}, err
+	}
+
+	if userdata[:len("#cloud-config")] != "#cloud-config" {
+		// add the header if it's missing. When parsing multipart MIME, we get the type
+		// of the userdata from the Content-Type header, so we don't require that the body contain
+		// the header, but it simplifies our lives if we add it here, as there are functions that look
+		// for it in other parts of the codebase.
+		userdata += "#cloud-config\n\n"
```

Comment:
Should this rather be:
```suggestion
		userdata = "#cloud-config\n\n"+userdata
```